expertsystem/data.py

```diff
@@ -0,0 +1,191 @@
+        if value is not None and value > self.value:
+            raise ValueError(
+                "Cannot set minimal value higher than value"
+                f" ({value} > {self.value})"
+            )
+        self.__min = value
+
+    @property
+    def max_value(self) -> Optional[float]:
+        return self.__max
+
+    @max_value.setter
+    def max_value(self, value: Optional[float]) -> None:
+        if value is not None and value < self.value:
+            raise ValueError(
+                "Cannot set maximal value lower than value"
+                f" ({value} < {self.value})"
+            )
+        self.__max = value
+
+
+class Parity:
+    """Safe, immutable data container for parity."""
+
+    def __init__(self, value: Union[float, int, str]) -> None:
+        value = float(value)
+        if value not in [-1.0, +1.0]:
+            raise ValueError("Parity can only be +1 or -1")
+        self.__value: int = int(value)
+
+    def __eq__(self, other: object) -> bool:
+        return self.__value == other
+
+    def __int__(self) -> int:
+        return self.value
+
+    def __repr__(self) -> str:
+        return (
+            f'<{self.__class__.__name__} {"+1" if self.__value > 0 else "-1"}>'
+        )
+
+    @property
+    def value(self) -> int:
+        return self.__value
+
+
+class Spin:
+    """Safe, immutable data container for (iso)spin."""
+
+    def __init__(self, magnitude: float, projection: float) -> None:
+        if abs(projection) > magnitude:
+            raise ValueError(
+                "Spin projection cannot be larger than its magnitude"
+            )
+        self.__magnitude = float(magnitude)
+        self.__projection = float(projection)
+
+    def __eq__(self, other: object) -> bool:
+        return self.__magnitude == other
+
+    def __float__(self) -> float:
+        return self.__magnitude
+
+    def __repr__(self) -> str:
+        return f"<{self.__class__.__name__} {self.__magnitude, self.__projection}>"
+
+    @property
+    def magnitude(self) -> float:
+        return self.__magnitude
+
+    @property
+    def projection(self) -> float:
+        return self.__projection
+
+
+class MeasuredValue(NamedTuple):
+    """Value with (optional) uncertainty, as reported by a measurement."""
+
+    value: float
+    uncertainty: Optional[float] = None
+
+    def __eq__(self, other: object) -> bool:
+        return self.value == other
+
+    def __float__(self) -> float:
+        return self.value
+
+    def __repr__(self) -> str:
+        if self.uncertainty is None:
+            return str(self.value)
+        return f"{self.value} ± {self.uncertainty}"
+
+
+class Particle(NamedTuple):
+    """Immutable data container for particle info."""
+
+    name: str
+    pid: int
+    charge: float
+    spin: float
+    mass: MeasuredValue
+    strange: int = 0
+    charm: int = 0
+    bottom: int = 0
+    top: int = 0
+    baryon: int = 0
+    ln_e: int = 0
+    ln_mu: int = 0
+    ln_tau: int = 0
```

Comment:
We should think about the names. I would tend to use the official names, even if they are slightly longer

Reply:
You mean for instance `lepton_number_e`?

Reply:
yeah, `strangeness`, `baryon_number` etc

Reply:
Ah ok. Yes for `strange` I had in mind something like `j_psi.strange = 0`. But I guess `j_psi.strangeness = 0` is more natural indeed... Only the `lepton_number_e`... meh, not sure if I would prefer that over `ln_e`.